### ISSUE TITLE
CuSPARSE support

### ISSUE THREAD

**Issue by mhobby** [tensorflow/tensorflow#5193]
For some machine learning /algorithm problems, we require to multiply matrices with vectors. In some cases - especially those associated with large graph analytics, the matrices are sparse. Given the major limitation with GPUs is the limited GPU memory, representing the matrix in sparse format is vital.

To this end, the CUSPARSE libraries (http://docs.nvidia.com/cuda/cusparse/) provide a suitable object model and various functions for sparse matrix/vector operations.

I would therefore like to request a basic CUSPARSE implementation. 
e.g. - 
- CSR matrix representation - (see http://docs.nvidia.com/cuda/cusparse/#compressed-sparse-row-format-csr)
- basic matrix-vector multiplication for CSR representations (see http://docs.nvidia.com/cuda/cusparse/#cusparse-lt-t-gt-csrmv)

NOTE: Only file GitHub issues for bugs and feature requests.  All other topics will be closed.

For general support from the community, see [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow).
To make bugs and feature requests more easy to find and organize, we close issues that are deemed
out of scope for GitHub Issues and point people to StackOverflow.

For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.
### Environment info

Operating System:

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

If installed from binary pip package, provide:
1. A link to the pip package you installed:
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.

If installed from source, provide 
1. The commit hash (`git rev-parse HEAD`)
2. The output of `bazel version`
### Logs or other output that would be helpful

(If logs are large, please upload as attachment or provide link).

**Comment by drpngx** [comment]
Yes, that would be great. Is that something you would be willing to contribute?

**Comment by drpngx** [comment]
@martinwicke @vincentvanhoucke for comments

**Comment by martinwicke** [comment]
Maybe interested parties: @rmlarsen @aselle

**Comment by vincentvanhoucke** [comment]
We have sparse matrix support here:
https://www.tensorflow.org/versions/r0.11/api_docs/python/sparse_ops.html
I don't know what the interop story with CSR representations in general and CUSPARSE in particular might look like.

**Comment by aselle** [comment]
I'm definitely interested in more support for accelerated sparse matrix formats like CSR. It will likely not appear soon, and we still are thinking about design and impact of such a feature. Any further information on use case and your experience with our existing index based sparse matrix format would be very helpful.

**Comment by aselle** [comment]
@mhobby, Did you manage to work around this in anyway, or do you still need this support?

**Comment by mhobby** [comment]
HI @aselle - we found a work around by using Julia which has direct implementation/wrapper of the CSR routines in CUDA.
Has there been any further progress in tensorflow for this?

**Comment by aselle** [comment]
There really has not been any application that has driven that forward for us as of yet. So, for now I am marking this contributions welcome, although there would likely need to be a pretty good design document to make a case for way to integrating this into TensorFlow elegantly to go with any effort to solve the feature.

**Comment by qingyangqing** [comment]
Hi @aselle , I am also looking forward to having the support from tf to accelerate sparse matrix computations. sparse_matmul op is heavily used in my case, but this sparse op cannot compete with the normal matmul op on CPU with SSE support, not to say on GPU. I also find sparse_matmul op cannot run on GPU. 

I can say my motivation to have sparse matrix support. For neural network, many papers have proven neuron weights can be pruned, thus the matmul turns to be sparse matrix multiply, which gives speedup space and saves energy consumption.

**Comment by qingyangqing** [comment]
Hi @aselle, could you tell me more about how to contribute ? After searching TF, Mxnet and Caffe,  I find TF has done the most on sparse matrix support. Although TF support sparse tensors, but it still lacks optimization and acceleration for sparse matrix ops. So I hope to do something to improve sparse matrix computations, which will also benefit my daytime work. If someone has already started on this issue,   please let me known that I can join. 
Looking forward to your feedback.

**Comment by ebrevdo** [comment]
I'm looking into this, adding cusparse support is nontrivial.  I wouldn't expect anything until August at the earliest.  For now we go have a GPU accelerated sparse_tensor_dense_matmul.

**Comment by ebrevdo** [comment]
(it's not trivial because the primitives don't have batched computation
support)

**Comment by xysmlx** [comment]
Hi, Qing @qingyangqing , in the density of pruned neural networks (about 10%-30%), sparse-dense matrix multiplication **cannot** outperform dense-dense matrix multiplication both in GPU and CPU. Although the sparse method reduces lots of computation, the utilization of CPU/GPU is much lower than that of dense-dense matrix multiplication. According to experiments, GEMM can achieve up to 90%+ peak flops and CSRMM can only achieve <5% peak flops in GPUs.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity, but please let me know if I'm mistaken. Thanks!

**Comment by EliasHasle** [comment]
OpenAI has recently published a library for GPU-accelerated block-sparse operations: https://github.com/openai/blocksparse

The lowest allowed block size is 8. Block size 1 would make it properly sparse, but I guess that was too inefficient and not necessary to achieve benefits.
